### PR TITLE
[Feat][Dashboard] Add global UTC timezone button in navbar with local storage

### LINKED ISSUE #47762: [Observability] Ability to show timestamps in UTC
### Description

A global button which will allow users to modify the timezone dates are shown. Ideally the user's setting is saved in local storage so they do not need to adjust every time.

May not be possible to adjust log line entries, but at least for some UI rendered dates we should be able to do it.

### Use case

_No response_

### PR DESCRIPTION
**Pull request by nadongjun**
## Why are these changes needed?

- #47762 

### Changes Included
Timezone Selection Button: Added a button at the top of the UI that lets users choose their preferred timezone (UTC). This selection will update displayed dates accordingly in supported sections.

Persistent Settings: The selected timezone is saved to the browser’s local storage, ensuring that the setting remains consistent upon page reloads.

### Supported Sections:

Three main areas were considered for timezone adjustments:

- [x] 1. Grafana-Related Data: The selected timezone setting is now applied using the timezone parameter to set UTC in the URL. https://grafana.com/blog/2024/10/23/grafana-11.3-release-all-the-new-features/
- [x] 2. Tables and Details for Jobs, Nodes, Actors, and Workers: The selected timezone setting is fully applied to these sections.
- [ ] 3. Log Data: Currently unable to support timezone adjustments due to technical limitations.

Scope: Given current limitations, only the Log Data section (Point 3 above) does not support timezone adjustments, while Grafana-Related Data and the Jobs, Nodes, Actors, and Workers sections are fully updated to reflect the timezone selection.

### Result:
image

## Related issue number

Closes #47762

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by nadongjun** [comment]
@alanwguo Hi,

Currently, the timezone adjustment is only applied to the detail sections for Jobs, Nodes, Actors, and Workers that use the `formatByteAmount` function. However, I'm having difficulty supporting timezone adjustments for Grafana-related data (due to API limitations)

If you have any suggestions or alternative methods to handle timezone adjustments in these areas, I’d appreciate your insights! Thanks!

**Comment by nadongjun** [comment]
https://grafana.com/blog/2024/10/23/grafana-11.3-release-all-the-new-features/
- [example](https://play.grafana.org/d/T512JVH7z/loki-nginx-service-mesh-json-version?from=2024-11-03T21:14:29.874Z&to=2024-11-04T03:14:29.874Z&var-datasource=ac4000ca-1959-45f5-aa45-2bd0898f7026&var-label_name=filename&var-label_value=%2Fvar%2Flog%2Fnginx%2Fjson_access.log&var-job=$__all&var-instance=$__all&viewPanel=panel-2&tz=America%2FCordoba&timezone=Asia/Seoul)

I just checked the Grafana 11.3 release, and it looks like timezone adjustments are now supported. I will make sure to incorporate this.

**Comment by nadongjun** [comment]
@alanwguo Could you take a look at this when you have a moment?

File: python/ray/dashboard/client/src/pages/layout/MainNavLayout.tsx [hunk 1]
```diff
@@ -108,6 +118,176 @@ const MainNavBar = () => {
...
+  useEffect(() => {
+    localStorage.setItem("timezone", timezone);
+  }, [timezone]);
+
+  const handleTimezoneChange = (event: SelectChangeEvent<string>) => {
+    setTimezone(event.target.value);
+    window.location.reload();
+  };
+
+  const timezones = [
```
Comment (alanwguo) [inline python/ray/dashboard/client/src/pages/layout/MainNavLayout.tsx:138]: where did this list come from? Does this cover most common time zones?

**Review by alanwguo — COMMENTED** [review]
Thanks!

Can we add a description or tooltip near the timezone selector that explains that timezone of logs may not respect this value?

File: python/ray/dashboard/client/src/pages/layout/MainNavLayout.tsx (hunk 1 above)
Comment (nadongjun) [reply python/ray/dashboard/client/src/pages/layout/MainNavLayout.tsx:138]: This list was exported from GitHub's Timezone selection options.
image

**Comment by nadongjun** [comment]
@alanwguo I added a tooltip with the message "The timezone of logs may not match this selection." Thanks for the review!

**Comment by alanwguo** [comment]
@scottsun94, plz review as well.

A couple of suggestions. When the dropdown is collapsed, can we make the contents of the dropdown slimmer. Like "GMT-8" instead of the full name.

Second. Can we detect the timezone used by the head node and set that as the default for the UI? I think in many cases, the user's time zone won't match the system's time zone which is often UTC for cloud-based instances. By setting the default to the head node, at least the logs and the UI will match timezones.

**Comment by nadongjun** [comment]
@alanwguo Thanks for the review,

The first suggestion sounds doable.

For the second, would it work to set the default timezone to the timezone where the dashboard server (React server) is running?

Also, is there a separate API to fetch the UTC timezone of the head node?

**Comment by scottsun94** [comment]
Here is a great article about the timezone UX: https://www.linkedin.com/pulse/designers-guide-time-zone-selection-ux-vitaly-friedman/.

Grafana seems a great example to learn from:
1. Have suggested time zones at the top. We can consider  `Ray cluster time` (or `Dashboard server time`), `Browser time`, `UTC`
2. Detect and show `dashboard/cluster time` by default
3. show city, country (dimmed), timezone names (dimmed) and allow users to search by them. Maybe just follow what Grafana has?
- if it's easy, we can try grouping by continents & sort by alphabetically as well
Screenshot 2024-11-06 at 6 35 01 PM

**Comment by nadongjun** [comment]
@scottsun94 Thank you for sharing the great article. I will refer to it and incorporate the suggestions accordingly.

**Comment by alanwguo** [comment]
> @alanwguo Thanks for the review,
> 
> The first suggestion sounds doable.
> 
> For the second, would it work to set the default timezone to the timezone where the dashboard server (React server) is running?
> 
> Also, is there a separate API to fetch the UTC timezone of the head node?

There is no API for this for now, but you can add a new API to `http_server_head` and call it and add the timezone to GlobalContext in `App.tsx`

**Comment by nadongjun** [comment]
@alanwguo 

Thank you for the guidance! Here's how I plan to proceed with the additional changes:

1. Refactor the timezone select box into a separate component and add a search feature for better usability.
2. Add a new API to http_server_head to fetch the server-side timezone and set it as the default timezone.

Please let me know if you have any additional feedback.

**Comment by scottsun94** [comment]
@nadongjun after the UI change is done, can you share a short walkthrough video or screenshots? Would love to take another look.

**Comment by nadongjun** [comment]
@scottsun94 Sure, I am currently working on the UI changes first. Once the updates are complete, I will share a video walkthrough and tag you.

**Comment by nadongjun** [comment]
> @nadongjun after the UI change is done, can you share a short walkthrough video or screenshots? Would love to take another look.

I've implemented the changes using the [react-select library](https://react-select.com/home).

https://github.com/user-attachments/assets/06076efa-1712-4217-a797-0a3ff506e1b5

**Comment by alanwguo** [comment]
@nadongjun have you tried implementing with https://mui.com/material-ui/react-autocomplete/ ?

Then we can use the existing MUI library and don't have to take on a new dependency

**Comment by scottsun94** [comment]
Played with the design a bit. Wondering if we should:
1. Show the `GMT+xx:xx` to make it conciser so that you can reduce the width of the box
2. add a divider between (browser time, coordinated universal time, etc.) and the rest of the time zones
3. adjust the format for each row to something like `{Continent}/{City}  {Country}, {timezone name (if exists)}      {offset from GMT}  `
4. sort by Continent and city


Screenshot 2024-11-08 at 11 19 41 AM

**Comment by nadongjun** [comment]
https://github.com/user-attachments/assets/380c854c-be7a-4e54-9187-84c6723451ba
- The timezone list content and the System category are currently being updated.

@alanwguo Thank you for the suggestion to use Autocomplete from MUI. I've updated the implementation to use the existing MUI library, so no new dependencies are added.

@scottsun94 
Regarding the design changes:

- I’ve added sorting and categorized time zones by continent, as I thought it would be helpful to display time zones grouped by continent.
- Each row now displays in the format {Continent}/{City} {Country} {GMT Offset}.
- The input box width has also been reduced for a more concise display.

Please let me know if you think any changes are needed. Thanks again for the helpful input!

**Comment by scottsun94** [comment]
Thanks for the quick turnaround. Nice work!

A few small visual nits before we are good to go:

1. Can you align the box with the rest of the UI on the right? We should have a 24px padding to the right.
2. Can we change the font color to #5F6469 (secondary) when it's not hovered/selected? And when it's selected, can you use #8C9196 (tertiary)? We use these two shades else where. I forgot if we have existing variables for them. You can take a look.
3. Can you make sure the border color of the box is the same as other borders in the UI?

Screenshot 2024-11-08 at 9 40 02 PM

4. Can we align the group title and the timezones on the left? Don't see an obvious value to add indent here?
5. The spacing between the group title and the rest of the timezones can be smaller. Can be the same as the spacing between the time zones

Screenshot 2024-11-08 at 9 40 17 PM

6. Can we add dividers between different system/continents like this? 
7. Can we use the following font colors for different texts?
- group titles: continent/system: #5F6469 (secondary)
- {Continent}/{city}, GMT offsets: #000 (primary)
- {country}: #8C9196 (tertiary)

Screenshot 2024-11-08 at 9 57 14 PM

**Comment by nadongjun** [comment]
@alanwguo 

I’ve added the tzlocal library as a dependency in the dashboard server to read the system timezone. Is this approach okay with you? Let me know if you have any concerns.

@scottsun94 

Thank you so much for the quick response and helpful feedback!

I really appreciate the UI suggestions you've provided. However, I'm new to UI-related coding, and I'm a bit unsure how to implement all the changes you suggested. Would it be possible for you to help me with the code for these adjustments?

**Comment by scottsun94** [comment]
> @alanwguo
> 
> I’ve added the tzlocal library as a dependency in the dashboard server to read the system timezone. Is this approach okay with you? Let me know if you have any concerns.
> 
> @scottsun94
> 
> Thank you so much for the quick response and helpful feedback!
> 
> I really appreciate the UI suggestions you've provided. However, I'm new to UI-related coding, and I'm a bit unsure how to implement all the changes you suggested. Would it be possible for you to help me with the code for these adjustments?

Thanks for the note! I'll leave the decision about the UI part to @alanwguo

File: python/ray/dashboard/client/src/components/SearchComponent.tsx [hunk 2]
```diff
@@ -77,3 +82,927 @@ export const SearchSelect = ({
...
+  }, [serverTimeZoneLoaded, serverTimeZone]);
+
+  useEffect(() => {
+    if (serverTimeZoneLoaded) {
+      localStorage.setItem("timezone", timezone);
+    }
+  }, [timezone, serverTimeZoneLoaded]);
+
+  const handleTimezoneChange = (value: string) => {
+    setTimezone(value);
```
Comment (alanwguo) [inline python/ray/dashboard/client/src/components/SearchComponent.tsx:919]: let's just set the localStorage value here instead of the useEffect.

There seems to be a race condition. The window might reload before the effect runs.

For the case where the timezone is set initially by the server, there shouldn't be a need to set the local storage, we can say local storage is only used for user manually set values. Your existing logic of use local storage and then fall-back to server time zone is the right behavior.

**Review by alanwguo — COMMENTED** [review]
I can take a look at the CSS issues. I'll add a commit to this branch after you address my other comments.

**Comment by nadongjun** [comment]
> I can take a look at the CSS issues. I'll add a commit to this branch after you address my other comments.

I’ll address your comments and let you know once the changes are committed. Appreciate your help with the CSS issues.

**Comment by nadongjun** [comment]
https://github.com/user-attachments/assets/f2fa8bbb-171b-433a-8870-c4f2e0bd0442

@alanwguo 

As per your review, I have made changes to avoid using external libraries. The dashboard server now maintains a single list of timezone names corresponding to each offset and returns the appropriate timezone name for a given offset via the /timezone API.

Additionally, instead of using useEffect, the data is now directly saved to localStorage, and the timezone name retrieved from the /timezone API is not stored in localStorage.

For UI adjustments, the Dashboard Server Timezone will display Asia/Irkutsk, which is the representative timezone for +09:00 on the server side, as seen in the demo video. It might be better to hide the value for the system timezone in this case.

Please feel free to share any additional feedback regarding the changes.

**Comment by scottsun94** [comment]
> For UI adjustments, the Dashboard Server Timezone will display Asia/Irkutsk, which is the representative timezone for +09:00 on the server side, as seen in the demo video. It might be better to hide the value for the system timezone in this case.

Yeah. agree.

Another possible small enhancement: maybe use a banner to make it more prominent that logs may not match the selection in the dropdown itself.
Screenshot 2024-11-15 at 12 04 26 PM

File: python/requirements.txt [hunk 3]
```diff
@@ -60,4 +60,4 @@ pandas>=1.3
 pydantic!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.*,!=2.4.*,<3  # Serve users can use pydantic<2
 py-spy>=0.2.0
 memray; sys_platform != "win32" # memray is not supported on Windows
-pyOpenSSL
+pyOpenSSL
```
Comment (alanwguo) [inline python/requirements.txt:63]: revert the changes to this file plz.

File: python/ray/dashboard/client/src/components/SearchComponent.tsx [hunk 4]
```diff
@@ -77,3 +83,910 @@ export const SearchSelect = ({
     </TextField>
   );
 };
+
+export const SearchTimezone = ({
```
Comment (alanwguo) [inline python/ray/dashboard/client/src/components/SearchComponent.tsx:87]: Let's move this to a separate file `SearchTimezone.tsx`

File: python/ray/dashboard/client/src/App.tsx [hunk 5]
```diff
@@ -91,6 +95,12 @@ export type GlobalContextType = {
    * The name of the current selected datasource.
    */
   dashboardDatasource: string | undefined;
+  /**
+   * The name of the current server-side timezone.
+   */
+  serverTimeZone: string | undefined;
+
+  serverTimeZoneLoaded: boolean | undefined;
```
Comment (alanwguo) [inline python/ray/dashboard/client/src/App.tsx:103]: let's combine these into a single variable.

`serverTimeZone` will either be undefined if it's unloaded. null, if we couldn't fetch it, or the value from the server.

File: python/ray/dashboard/http_server_head.py [hunk 6]
```diff
@@ -139,6 +140,68 @@ async def get_favicon(self, req) -> aiohttp.web.FileResponse:
             )
         )
 
+    def get_gmt_offset(self):
+        current_tz = datetime.now().astimezone().tzinfo
+        offset = current_tz.utcoffset(None)
+        hours, remainder = divmod(offset.total_seconds(), 3600)
+        minutes = remainder // 60
+        sign = "+" if hours >= 0 else "-"
+        return f"GMT{sign}{abs(int(hours)):02d}:{abs(int(minutes)):02d}"
```
Comment (alanwguo) [inline python/ray/dashboard/http_server_head.py:149]: let's move this and most of the implementation of `get_timezone` into a separate file. SOmething like `timezone_utils.py`

File: python/ray/dashboard/http_server_head.py [hunk 7]
```diff
@@ -139,6 +140,68 @@ async def get_favicon(self, req) -> aiohttp.web.FileResponse:
...
+
+    @routes.get("/timezone")
+    async def get_timezone(self, req) -> aiohttp.web.Response:
+        timezones = [
```
Comment (alanwguo) [inline python/ray/dashboard/http_server_head.py:153]: after moving this to a new file, let's up-level this to be a file-level const.

File: python/ray/dashboard/http_server_head.py [hunk 8]
```diff
@@ -139,6 +140,68 @@ async def get_favicon(self, req) -> aiohttp.web.FileResponse:
...
+            {"utc": "GMT-12:00", "value": "Etc/GMT+12"},
```
Comment (alanwguo) [inline python/ray/dashboard/http_server_head.py:154]: instead of `utc` the key should be `offset`

File: python/ray/dashboard/http_server_head.py [hunk 9]
```diff
@@ -139,6 +140,68 @@ async def get_favicon(self, req) -> aiohttp.web.FileResponse:
...
+            {"utc": "GMT+13:00", "value": "Pacific/Tongatapu"},
+        ]
+        try:
+            current_offset = self.get_gmt_offset()
+            current_timezone = next(
+                (tz["value"] for tz in timezones if tz["utc"] == current_offset), None
+            )
+
+            if current_timezone:
+                return aiohttp.web.Response(text=str(current_timezone))
```
Comment (alanwguo) [inline python/ray/dashboard/http_server_head.py:195]: instead of returning raw text, let's return a simple json dictionary so we may add additional fields in the future.

**Review by alanwguo — COMMENTED** [review]
Thanks! This is looking close. Just some minor stylistic changes.

Please just address my comments, I'll work on any follow-up UI changes including the banner @scottsun94 suggested.

**Comment by nadongjun** [comment]
@alanwguo

Thank you for your feedback. I’ve incorporated your suggestions.

- Reverted and recommitted the changes to python/requirements.txt.
- Updated the raw string in the dashboard API (/timezone) response to use JSON format ({offset, value}).
- Refactored timezone-related logic in the dashboard API into a separate function and moved the client-side timezone list to a dedicated file.

Let me know if there’s anything else to improve!

File: python/ray/dashboard/client/src/components/SearchComponent.tsx [hunk 10]
```diff
@@ -77,3 +84,115 @@ export const SearchSelect = ({
...
+  serverTimeZone,
+  serverTimeZoneLoaded,
+}: {
+  serverTimeZone?: string | null;
+  serverTimeZoneLoaded?: boolean;
+}) => {
+  const [timezone, setTimezone] = useState<string>("");
+
+  useEffect(() => {
+    if (serverTimeZoneLoaded) {
```
Comment (alanwguo) [inline python/ray/dashboard/client/src/components/SearchComponent.tsx:98]: can we delete `serverTimeZoneLoaded` now that `serverTimeZone` has the `null` and `undefined` case?

**Review by alanwguo — COMMENTED** [review]
a couple more comments for a bit cleaner code.

**Comment by nadongjun** [comment]
> a couple more comments for a bit cleaner code.

Thank you for your feedback. I've addressed the review comments and made the following changes:

- Removed serverTimeZoneLoaded and modified the code to detect loading by checking if serverTimeZone is undefined.
- Fixed variable typos throughout the code.
- Updated the TimeZoneInfo object to have offset and value as strict string types
- (Additional change) Modified the Autocomplete options to calculate the browser's UTC offset using new Date() instead of searching from a predefined list.

**Comment by alanwguo** [comment]
Screenshot 2024-11-19 at 1 35 40 PM

Okay I pushed a commit that updates the styling. I wasn't able to get a banner to show up in the Autocomplete options list but I think i covered the other style changes.

I also refactored the code slightly to move the currentTimeZone logic into the GlobalContext since it is re-used in multiple places.

I updated the default time for the Overview page for now-1h instead of now-30m. For some really weird reason, the timezone query param doesn't work when the its 30 minutes, but works with every other value?

**Review by alanwguo — APPROVED** [review]

**Comment by scottsun94** [comment]
> Okay I pushed a commit that updates the styling. I wasn't able to get a banner to show up in the Autocomplete options list but I think i covered the other style changes.

So the message about the log timezone will still show up like this?

Screenshot 2024-11-19 at 1 50 56 PM

**Comment by alanwguo** [comment]
> > Okay I pushed a commit that updates the styling. I wasn't able to get a banner to show up in the Autocomplete options list but I think i covered the other style changes.
> 
> So the message about the log timezone will still show up like this?
> 
> Screenshot 2024-11-19 at 1 50 56 PM

yeah

**Comment by scottsun94** [comment]
> > > Okay I pushed a commit that updates the styling. I wasn't able to get a banner to show up in the Autocomplete options list but I think i covered the other style changes.
> > 
> > 
> > So the message about the log timezone will still show up like this?
> > Screenshot 2024-11-19 at 1 50 56 PM
> 
> yeah

Okay. SGTM

**Comment by nadongjun** [comment]
Could anyone merge this PR?

**Comment by scottsun94** [comment]
@nadongjun 

I realize that this timezone picker also doesn't affect several other UI elements:
- embedded grafana graphs
- the timestamps of the events in the overview page

Screenshot 2025-02-04 at 10 59 38 AM


The default timezone seems to be dashboard serve's timezone? Shall we actually use browser timezone instead?

**Comment by nadongjun** [comment]
@scottsun94 Thank you for checking this. I will verify whether the mentioned UI can have the timezone applied. 

Given the current structure, if the Timezone Selector is not used, the default timezone follows the Ray dashboard server's timezone. I believe the server timezone is more appropriate in this case.

**Comment by scottsun94** [comment]
>  I will verify whether the mentioned UI can have the timezone applied.

Thanks!

> Given the current structure, if the Timezone Selector is not used, the default timezone follows the Ray dashboard server's timezone. I believe the server timezone is more appropriate in this case.

Okay. Let's see how people feel about it.
